elastic/kibana
# [Security Solution] [Detection Engine] Fix `rule_exception` cypress test flakiness

## Summary

- Fixes https://github.com/elastic/kibana/issues/159341 

**Steps to insure the flakiness** 

- Run the test through Flaky test runner by modifying the `x-pack/plugins/security_solution/package.json` to only run the `Rule_exceptions` as the following =>  `./cypress/e2e/exceptions/**/rule_exception.cy.ts` 
- Run the test locally multiple times by wrapping the test suite by 
 `Cypress._.times(50, () => {});`

**Failing steps**

- The test only failed in the step mentioned when executed through the Flaky test runner.

```
  1) Rule Exceptions workflows from Alert
  --
    | Should create a Rule exception item from the alert actions overflow menu and close all matching alerts:
    | AssertionError: Timed out retrying after 150000ms: Expected to find element: `[data-test-subj="alertsStateTableEmptyState"]`, but never found it.
```

- By executing it locally multiple times, the test failed at different steps on each occasion. 
  
   - Immediately after creating an Exception with the Closing all matching alerts option, the Alert table was not found empty, indicating a failure.
![image](https://github.com/elastic/kibana/assets/12671903/492aa552-f721-493b-b0b7-fa3e91dc4e63)

   -  After removing the exception, the expected number of alerts was not found in the Alerts table, resulting in a failure.
    ![image](https://github.com/elastic/kibana/assets/12671903/c291d630-8775-4a5b-9a6b-feb30a169757)

   - Unrelated to this test, but concerning exceptions, the failure occurred when attempting to click on "Submit exception" due to an error toast obstructing the button, similar to the 2nd screenshot provided below.
    ![image](https://github.com/elastic/kibana/assets/12671903/d117df13-d521-4ead-8a71-e7a6bd1585d8)
     ![image](https://github.com/elastic/kibana/assets/12671903/758cbb3e-9b95-439a-a5d1-49de8d84518e)

**Steps to mitigate test flakiness:**

- **Splitting Test Cases:** The rule_exceptions test currently covers multiple areas using different Rule types. To improve stability, we should divide it into separate test files. For example, `closing_all_matching_alerts.cy.ts` and `auto_populate_with_alert_data.cy.ts`. Additionally, we should ensure that each test file cleans up any previous data to maintain test isolation.

- **Handling Async Calls:** Test flakiness can occur when Cypress executes faster than async calls, leading to errors when alerts don't move to the closed tab in time. To address this, we should avoid immediate validation after creating an exception. Instead, refresh the `Alerts table` to allow time for the alerts to be moved to the appropriate tab before asserting the conditions.

   For example ` waitForAlerts();` method

- **Avoiding Specific Alert Counts:** Relying on a specific number of generated or closed alerts for validation is not a robust approach. During test execution, the Rule might generate more alerts than expected, leading to test failures. To overcome this, we should focus on validating the expected behavior rather than a fixed alert count.
  
     For example
      ```
          goToClosedAlertsOnRuleDetailsPage();  
          cy.get(ALERTS_COUNT).should('exist');
      ```

- In the test, when utilizing `cy.task('esArchiverLoad', 'exceptions');`, it can be beneficial to check if the exception list is not present in the environment by calling `cy.task('esArchiverUnload', 'exceptions');` beforehand. This cleanup step ensures that any existing exception list is removed, mitigating the occurrence of the following error: `resource_already_exists_exception: index [.lists-default-000001/U_3_TShzRY-WZD_XqhSBnw] already exists (400).`

- It does not provide a value at the end of the `closing_all_matching_alerts.cy.ts` test to verify if alerts are generated again after the `Exception` that closes all matching alerts is deleted because the test already begins with the rule generating alerts until the exception is created. Therefore, this part has been excluded.  

By following these steps for the test, running the new files multiple times locally proved to be beneficial:

![image](https://github.com/elastic/kibana/assets/12671903/8a0250a0-274f-442b-b94c-86e4c815afd1)
 successfully 
![image](https://github.com/elastic/kibana/assets/12671903/4e9562f3-506e-4f13-ab1f-c084d2613eb8)
